test/shell/test_scalafmt.sh

```diff
@@ -0,0 +1,58 @@
+# shellcheck source=./test_runner.sh
+dir=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+. "${dir}"/test_runner.sh
+. "${dir}"/test_helper.sh
+runner=$(get_test_runner "${1:-local}")
+
+run_non_default_formatting() {
+  set +e
+
+  FILE_PATH="$( dirname $( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd ) )"/scalafmt
+  RULE_TYPE=$1
+  FILENAME="formatted"
+  if [[ $RULE_TYPE = test ]]; then
+    FILENAME="formatted-test"
+  fi
+    
+  bazel run //test/scalafmt:formatted-$RULE_TYPE.format-test
+  if [ $? -ne 0 ]; then
+    echo -e "${RED} formatted-$RULE_TYPE.format-test should be a formatted target. $NC"
+    exit 1
+  fi
+
+  bazel run //test/scalafmt:unformatted-$RULE_TYPE.format-test
+  if [ $? -eq 0 ]; then
+    echo -e "${RED} unformatted-$RULE_TYPE.format-test should be an unformatted target. $NC"
+    exit 1
+  fi
+
+  bazel run //test/scalafmt:unformatted-$RULE_TYPE.format
+  if [ $? -ne 0 ]; then
+    echo -e "${RED} unformatted-$RULE_TYPE.format should run formatting. $NC"
+    exit 1
+  fi
+
+  diff $FILE_PATH/un${FILENAME}.scala $FILE_PATH/${FILENAME}.scala
+  if [ $? -ne 0 ]; then
+    echo -e "${RED} un${FILENAME}.scala should be the same as ${FILENAME}.scala after formatting. $NC"
+    exit 1
+  fi
+
+  cp $FILE_PATH/un${FILENAME}.template.scala $FILE_PATH/un${FILENAME}.scala
```

Comment:
should we do this in the beginning of the test and then add the `un${FILENAME}.scala` to .gitignore?
What we'll get is that we don't need to keep the three of those checked in and also if the test fails you have the files to look at

Reply:
Let me think about a better way to test. Get back to you soon.

Reply:
I like your suggestion. But we have to keep `un${FILENAME}.scala` checked in otherwise `bazel build test/...` in `test_rules_scala.sh` will complain no such source files. Instead, we don't have to check in `un${FILENAME}.template.scala`.

Re-organize the test a bit. Please take a look again.